source/common/filter/tcp_proxy.cc

```diff
@@ -370,6 +383,28 @@ void TcpProxy::onUpstreamEvent(Network::ConnectionEvent event) {
     read_callbacks_->upstreamHost()->outlierDetector().putResult(
         Upstream::Outlier::Result::SUCCESS);
     onConnectionSuccess();
+
+    if (config_ != nullptr && config_->idleTimeout().valid()) {
+      idle_timer_ = read_callbacks_->connection().dispatcher().createTimer(
+          [this]() -> void { onIdleTimeout(); });
+      resetIdleTimer();
+      Network::BytesSentCb cb = [this](uint64_t) { resetIdleTimer(); };
+      read_callbacks_->connection().addBytesSentCallback(cb);
```

Comment:
I kind of wonder if we need to bother resetting the idle timer when we receive data at all, if ultimately all we care about is whether it gets sent in some period of time. If we remove the other cases we can avoid 1 of 2 timer recalculations for each side. Thoughts?

Reply:
I'm worried there will be corner cases where we expire a connection when we shouldn't have.

If the performance of resetting the timer is bad, that probably means we need a faster timers implementation (I haven't looked at how libevent implements timers).  

I think we should leave it as is until we've proved there's a problem here.

Reply:
It's not bad, but it's definitely not free. If you want to leave it that's fine, but I might add a perf TODO on thinking about removing it. I can't think of a corner case off the top of my head in which we would expire the connection.